### PR TITLE
Add branch cache support for GetTriples on population node

### PR DESCRIPTION
**Pull request by shifucun**

### CODE REVIEW

File: server/misc.go [hunk 1]
```diff
@@ -33,7 +33,7 @@ func (s *Server) Translate(ctx context.Context,
 	}
 
 	out := pb.TranslateResponse{}
-	mappings, err := translator.ParseMapping(in.GetSchemaMapping(), s.metadata.Bq)
+	mappings, err := translator.ParseMapping(in.GetSchemaMapping(), "bq")
```
Comment (pradh) [inline server/misc.go:36]: For my understanding, where is the DB used in the mapping (I tried following the func calls, it went multiple levels :)

File: server/triple.go [hunk 2]
```diff
@@ -179,6 +173,37 @@ func (s *Server) GetTriples(ctx context.Context, in *pb.GetTriplesRequest) (
...
+	if err != nil {
+		return nil, err
+	}
+	return &triples, nil
+}
+
+func convertPopTriples(dcid string, jsonRaw []byte) (interface{}, error) {
+	jsonVal := string(jsonRaw)
+	parts := strings.Split(jsonVal, "^")
+	if len(parts) == 0 || len(parts)%2 != 0 {
```
Comment (pradh) [inline server/triple.go:188]: nit: len(parts) % 2

?

File: server/triple.go [hunk 3]
```diff
@@ -75,6 +75,14 @@ func (s *Server) GetTriples(ctx context.Context, in *pb.GetTriplesRequest) (
 			resultsMap[dcid] = applyLimit(dcid, allTriplesCache[dcid].Triples, limit)
 		}
 	}
+	// Regular triples cache from memcache for population data
+	if len(popDcids) > 0 {
+		dataMap := s.memcache.ReadParallel(
+			buildTriplesKey(popDcids), convertTriplesCache)
```
Comment (pradh) [inline server/triple.go:81]: Do we have an inventory of which other cache types don't support branch cache?

File: main.go [hunk 4]
```diff
@@ -29,12 +29,13 @@ import (
...
-	btProject  = flag.String("bt_project", "", "GCP project containing the BigTable instance.")
-	btInstance = flag.String("bt_instance", "", "BigTable instance.")
-	projectID  = flag.String("project_id", "", "The cloud project to run the mixer instance.")
-	port       = flag.String("port", ":12345", "Port on which to run the server.")
+	bqDataset    = flag.String("bq_dataset", "", "DataCommons BigQuery dataset.")
+	btTable      = flag.String("bt_table", "", "DataCommons Bigtable table.")
+	btProject    = flag.String("bt_project", "", "GCP project containing the BigTable instance.")
+	btInstance   = flag.String("bt_instance", "", "BigTable instance.")
+	projectID    = flag.String("project_id", "", "The cloud project to run the mixer instance.")
+	branchFolder = flag.String("branch_folder", "", "The branch cache gcs folder.")
```
Comment (pradh) [inline main.go:37]: Is it just for testing?

File: main.go [hunk 5]
```diff
@@ -66,11 +67,15 @@ func main() {
...
 	}
 	// Memcache
-	branchCacheFolder, err := server.ReadBranchCacheFolder(
-		ctx, branchCacheBucket, branchCacheVersionFile)
-	if err != nil {
-		log.Fatalf("failed to read branch cache folder: %v", err)
+	branchCacheFolder := *branchFolder
+	if branchCacheFolder == "" {
+		branchCacheFolder, err = server.ReadBranchCacheFolder(
+			ctx, branchCacheBucket, branchCacheVersionFile)
```
Comment (pradh) [inline main.go:73]: Once we launch its going to be very tough to add anything to branch cache without risk of breaking prod. Not this PR, but soon, should we make separate paths for prod vs. staging and separate pubsub topics? Such that, the downstream pipeline writes to staging, and after validation, we'll have a script to: 1. copy over the branch cache folder file, 2. send prod notification.

File: examples/check_bt.sh [hunk 6]
```diff
@@ -0,0 +1,26 @@
...
+export BT=$(head -1 ../deployment/bigtable.txt)
+export PROJECT=google.com:datcom-store-dev
+echo $BT
+echo $PROJECT
+
+# Check get_stats cache
+echo "Get stats for d/f/country/AUT^count^measured^^^^Person^gender^Male"
+cbt -project=$PROJECT -instance prophet-cache read ${BT} prefix="d/f/country/AUT^count^measured^^^^Person^gender^Male"
+echo "Get stats for d/f/country/AUT^count^measured^Count_Person_25To64Years^^^Person^age^Years25To64^educationalAttainment^LessThanPrimaryEducation&PrimaryEducation&LowerSecondaryEducation"
+cbt -project=$PROJECT -instance prophet-cache read ${BT} prefix="d/f/country/AUT^count^measured^Count_Person_25To64Years^^^Person^age^Years25To64^educationalAttainment^LessThanPrimaryEducation&PrimaryEducation&LowerSecondaryEducation"
```
Comment (pradh) [inline examples/check_bt.sh:26]: This looks useful for debugging, should we make this an argument which defaults to this prefix?

File: main.go (hunk 4 above)
Comment (shifucun) [reply main.go:37]: Yes, this is for local testing with custom branch folder

File: main.go (hunk 5 above)
Comment (shifucun) [reply main.go:73]: yea, that sounds reasonable.

File: server/misc.go (hunk 1 above)
Comment (shifucun) [reply server/misc.go:36]: It's passed into here: https://github.com/datacommonsorg/mixer/blob/master/translator/mcf.go#L129

where it adds the bq to the column name prefix

File: server/triple.go (hunk 3 above)
Comment (shifucun) [reply server/triple.go:81]: This is something that need thorough examination. We have branch cache for many cache types but I am not entirely sure how are build, and how many overlap they have with base cache. This has indication of the merging. We should find a time later to go through each branch cache to figure out

1) if we need the branch cache
2) how it should be merged.

Comment (shifucun) [reply server/triple.go:81]: created https://github.com/datacommonsorg/mixer/issues/251 to track

File: examples/check_bt.sh (hunk 6 above)
Comment (shifucun) [reply examples/check_bt.sh:26]: I wan to throw a bunch of different bt cache keys in here for all cache types, that might need a lot arguments? what would be a good bash for this purpose?

Comment (shifucun) [reply examples/check_bt.sh]: ok, made the prefix as key and move the concrete ones as list comments above

Comment (pradh) [reply examples/check_bt.sh]: What you have works. Alternatively, we can accept input on stdin (refer: https://screenshot.googleplex.com/mhkM9hFpCDs), so they can do like below.

# echo "<KEY1>" | ./check_bt.sh

# cat key_file.txt
<KEY1>
<KEY2>
...
# cat key_file.txt | ./check_bt.sh

File: server/triple.go (hunk 3 above)
Comment (pradh) [reply server/triple.go:81]: Thanks, we can follow up there.

**Pull request merged**